cpp/src/sort/rank.cu

```diff
@@ -209,7 +209,7 @@ void rank_average(cudf::device_span<size_type const> group_keys,
   tie_break_ranks_transform<MinCount>(
     group_keys,
     cudf::detail::make_counting_transform_iterator(1,
-                                                   [] __device__(auto i) {
+                                                   [] __host__ __device__(size_type i) -> MinCount {
```

Comment:
Probably worth adding a link to #9703 on why this is being added.

Reply:
I'm a little confused by this. I'm confused, if the purpose of adding `__host__` is for return type deduction to work properly, doesn't explicitly specifying the return type render that superfluous? I'm probably missing something, but in case the two are indeed redundant I would prefer to remove the `__host__` qualifier and just keep the trailing return type.

Reply:
This should just be a functor, not a lambda.

Reply:
Why?

Reply:
The trailing return type did not have any effect on this issue. The issue does not seem to be return type deduction. Probably a compiler issue. NVBug 3451616 created with repro code. 
But this PR change is simply a workaround.

Reply:
OK just read the associated Slack thread. If the underlying issue is that we need to be able to query the return type and that's disallowed for device lambdas (although it seems we're not 100% sure exactly what the problem is?), a device functor seems like a more targeted solution than forcing host compilation. At minimum I agree with @davidwendt that a comment is necessary. Since it's just a (hopefully temporary) workaround I don't feel that strongly about which solution we go with.

Reply:
I agree. We should avoid `__host__ __device__` on code that is only needed on one or the other. Use a function object.

Reply:
Tested with device function object. Specifying trailing return type or return type in Device Function object makes the code work. But without return type (`auto`), it fails to compile (deduces return type to be void).

Reply:
Updated the code.